### PR TITLE
[GEOS-9380] Ignore remote WMS remote requests outside advertised bounds

### PR DESCRIPTION
**Pull request by imranrajjad**
An improvement in which user can skip remote WMS requests for cascaded layers if the requested area falls completely outside the native bounds advertised by remote layer. Option is available under Publishing tabs.

The unit test verifies this by creating a layer group of two cascaded layers (that do not completely overlap). Two mock requests validate that a GetMap completely outside the bounds of one of the layer, results in the remote call without the layer not in the requested bounds.

## Checklist

> Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

For all pull requests:

- [x] Confirm you have read the [contribution guidelines](https://github.com/geoserver/geoserver/blob/master/CONTRIBUTING.md) 
- [x] You have sent a Contribution Licence Agreement (CLA) as necessary (not required for small changes, e.g., fixing typos in documentation)
- [x] Make sure the first PR targets the master branch, eventual backports will be managed later. This can be ignored if the PR is fixing an issue that only happens in a specific branch, but not in newer ones.

The following are required only for core and extension modules (they are welcomed, but not required, for community modules):
- [x] There is a ticket in Jira describing the issue/improvement/feature (a notable exemptions is, changes not visible to end users)
- [x] PR for bug fixes and small new features are presented as a single commit
- [x] Commit message must be in the form "[GEOS-XYZW] Title of the Jira ticket" (export to XML in Jira generates the message in this exact form)
- [x] New unit tests have been added covering the changes
- [x] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [x] This PR passes the [QA checks](https://docs.geoserver.org/latest/en/developer/qa-guide/index.html) (QA checks results will be reported by travis-ci after opening this PR)
- [x] Commits changing the UI, existing user workflows, or adding new functionality, need to include documentation updates

Submitting the PR does not require you to check all items, but by the time it gets merged, they should be either satisfied or inapplicable.

### CODE REVIEW

File: src/main/src/main/java/org/geoserver/catalog/WMSLayerInfo.java [hunk 1]
```diff
@@ -54,4 +55,10 @@
     public Set<StyleInfo> getStyles();
 
     public StyleInfo getDefaultStyle();
+
+    public boolean isRespectMetadataBBox();
+
+    public void setRespectMetadataBBox(boolean respectMetadataBBox);
+
+    public boolean checkEnvelopOverLapWithNativeBounds(ReferencedEnvelope requestEnevelope);
```
Comment (aaime) [inline src/main/src/main/java/org/geoserver/catalog/WMSLayerInfo.java]: Nope, these are meant to be just be configuration carriers, no logic inside please.

File: src/web/wms/src/main/resources/GeoServerApplication.properties [hunk 2]
```diff
@@ -186,7 +186,7 @@ WMSLayerConfig.remoteStylesLabel = Default Remote Style
 WMSLayerConfig.remoteFormatsPaeletteLabel = Additional Remote Styles
 WMSLayerConfig.remoteFormatsDropDownLabel = Default Remote Format
 WMSLayerConfig.extraRemoteStylesLabel = Additional Remote Formats
-
+WMSLayerConfig.respectMetadataBBoxLabel = Do not send requests if requested Map Area is outside Layer bounds
```
Comment (aaime) [inline src/web/wms/src/main/resources/GeoServerApplication.properties]: "is outside of the Layer bounds" maybe?

File: src/wms/src/main/java/org/geoserver/wms/GetMap.java [hunk 3]
```diff
@@ -666,6 +666,18 @@ WebMap executeInternal(
                 WMSLayerInfo wmsLayer = (WMSLayerInfo) mapLayerInfo.getResource();
                 WebMapServer wms = wmsLayer.getStore().getWebMapServer(null);
                 Layer gt2Layer = wmsLayer.getWMSLayer(null);
+                if (wmsLayer.isRespectMetadataBBox()) {
+                    boolean isInsideBounnds =
+                            wmsLayer.checkEnvelopOverLapWithNativeBounds(
+                                    mapContent.getViewport().getBounds());
+                    if (!isInsideBounnds) {
+                        LOGGER.warning(
```
Comment (aaime) [inline src/wms/src/main/java/org/geoserver/wms/GetMap.java]: Definitely not something that should show up as warning, it's a normal behavior requested by the admin. I suggest replacing with a "FINE" log and guard it with a test on whether the logging level allows for FINE logging, or switch to a log with string + parameters (so that no string concatenation happens anyways).

File: src/wms/src/main/java/org/geoserver/wms/GetMap.java [hunk 4]
```diff
@@ -685,7 +697,7 @@ WebMap executeInternal(
                     WMSLayer Layer = null;
 
                     String style = request.getStyles().get(i).getName();
-                    style = (style == null) ? "" : style;
+                    style = (style == null) ? wmsLayer.getForcedRemoteStyle() : style;
```
Comment (aaime) [inline src/wms/src/main/java/org/geoserver/wms/GetMap.java]: Unrelated change?

File: src/wms/src/test/java/org/geoserver/wms/WMSCascadeTestSupport.java [hunk 5]
```diff
@@ -55,6 +58,7 @@ protected void onSetUp(SystemTestData testData) throws Exception {
         setupWMS130Layer();
         setupWMS110Layer();
         setupWMS110NfiLayer();
+        //  setupLayerGroup();
```
Comment (aaime) [inline src/wms/src/test/java/org/geoserver/wms/WMSCascadeTestSupport.java]: Don't keep commented out code sections.

File: src/wms/src/test/java/org/geoserver/wms/WMSCascadeTestSupport.java [hunk 6]
```diff
@@ -119,6 +123,54 @@ private void setupWMS130Layer() throws MalformedURLException, IOException {
...
+        group_lyr_2.setName("group_lyr_230");
+        group_lyr_2.reset();
+        group_lyr_2.setRespectMetadataBBox(true);
+        // group_lyr_1.setForcedRemoteStyle("line_group2");
+        getCatalog().add(group_lyr_2);
+        LayerInfo layer2 = cb.buildLayer(group_lyr_2);
+        getCatalog().add(layer2);
+
+        LayerGroupInfo group = getCatalog().getFactory().createLayerGroup();
+        // LayerGroupInfoImpl group = new LayerGroupInfoImpl();
```
Comment (aaime) [inline src/wms/src/test/java/org/geoserver/wms/WMSCascadeTestSupport.java]: Please don't leave commented out code sections.

File: src/wms/src/test/java/org/geoserver/wms/WMSCascadeTestSupport.java [hunk 7]
```diff
@@ -185,6 +237,57 @@ private void setupWMS110Layer() throws MalformedURLException, IOException {
...
+        // (LayerGroupInfoImpl) getCatalog().getFactory().createLayerGroup();
+
+        // group.setId("casc_group");
+        group.setName("cascaded_group");
+        group.getLayers().add(layer1);
+        group.getLayers().add(layer2);
+        try {
+            cb.calculateLayerGroupBounds(group);
+        } catch (Exception e) {
```
Comment (aaime) [inline src/wms/src/test/java/org/geoserver/wms/WMSCascadeTestSupport.java]: Just let it throw the exception then?

File: src/main/src/main/java/org/geoserver/catalog/WMSLayerInfo.java (hunk 1 above)
Comment (imranrajjad) [reply src/main/src/main/java/org/geoserver/catalog/WMSLayerInfo.java]: understood will move them out

File: src/web/wms/src/main/resources/GeoServerApplication.properties (hunk 2 above)
Comment (imranrajjad) [reply src/web/wms/src/main/resources/GeoServerApplication.properties]: sure, will fix

File: src/wms/src/main/java/org/geoserver/wms/GetMap.java (hunk 3 above)
Comment (imranrajjad) [reply src/wms/src/main/java/org/geoserver/wms/GetMap.java]: understood will do

File: src/wms/src/main/java/org/geoserver/wms/GetMap.java (hunk 4 above)
Comment (imranrajjad) [reply src/wms/src/main/java/org/geoserver/wms/GetMap.java]: yes, will remove

File: src/wms/src/test/java/org/geoserver/wms/WMSCascadeTestSupport.java (hunk 5 above)
Comment (imranrajjad) [reply src/wms/src/test/java/org/geoserver/wms/WMSCascadeTestSupport.java]: will remove

File: src/wms/src/test/java/org/geoserver/wms/WMSCascadeTestSupport.java (hunk 7 above)
Comment (imranrajjad) [reply src/wms/src/test/java/org/geoserver/wms/WMSCascadeTestSupport.java]: fail() is called in next line

File: src/wms/src/test/java/org/geoserver/wms/WMSCascadeTestSupport.java (hunk 6 above)
Comment (imranrajjad) [reply src/wms/src/test/java/org/geoserver/wms/WMSCascadeTestSupport.java]: will remove

**Comment by imranrajjad** [comment]
@aaime review please

**Comment by imranrajjad** [comment]
@aime thank you for your review, I have applied the requested changes, please let me know if any more changes or explanation is required?

note; I am unable to re-produce the app-veyor test fail locally

**Comment by imranrajjad** [comment]
@aaime thank you for your review, I have applied the requested changes, please let me know if any more changes or explanation is required?

File: src/main/src/main/java/org/geoserver/catalog/WMSLayerInfo.java (hunk 1 above)
Comment (aaime) [reply src/main/src/main/java/org/geoserver/catalog/WMSLayerInfo.java]: A method in GetMap works fine.

File: src/wms/src/test/java/org/geoserver/wms/WMSCascadeTestSupport.java (hunk 7 above)
Comment (aaime) [reply src/wms/src/test/java/org/geoserver/wms/WMSCascadeTestSupport.java]: It's un-necessary and hides the actual stack trace, so no, it's not.

File: src/main/src/main/java/org/geoserver/catalog/impl/WMSLayerInfoImpl.java [hunk 8]
```diff
@@ -71,12 +73,15 @@ public void reset() {
 
         try {
             // read from cap doc
-            return getWMSLayer(null)
-                    .getStyles()
-                    .stream()
-                    .map(s -> s.getName())
-                    .collect(Collectors.toList());
-
+            List<String> remoteStyles = new ArrayList<String>();
```
Comment (aaime) [inline src/main/src/main/java/org/geoserver/catalog/impl/WMSLayerInfoImpl.java]: This change seems unrelated, mixup with other work?

File: src/wms/src/main/java/org/geoserver/wms/GetMap.java [hunk 9]
```diff
@@ -988,4 +1003,20 @@ protected GetMapOutputFormat getDelegate(final String outputFormat) throws Servi
         }
         return producer;
     }
+
+    private boolean checkEnvelopOverLapWithNativeBounds(
+            ReferencedEnvelope requestEnevelope, ReferencedEnvelope layerEnevelope) {
+
+        try {
+            // transofrm requested enevelope to resource`s native bounds
```
Comment (aaime) [inline src/wms/src/main/java/org/geoserver/wms/GetMap.java]: transofrm -> transform

File: src/main/src/main/java/org/geoserver/catalog/impl/WMSLayerInfoImpl.java (hunk 8 above)
Comment (imranrajjad) [reply src/main/src/main/java/org/geoserver/catalog/impl/WMSLayerInfoImpl.java]: will remove

File: src/wms/src/main/java/org/geoserver/wms/GetMap.java (hunk 9 above)
Comment (imranrajjad) [reply src/wms/src/main/java/org/geoserver/wms/GetMap.java]: will fix

**Comment by imranrajjad** [comment]
@aaime thank you for your review, I have applied the requested changes, please let me know if any more changes or explanation is required?

**Comment by nmco** [comment]
@imranrajjad can you fix the conflicts?

**Comment by imranrajjad** [comment]
@nmco conflicts fixed

**Comment by imranrajjad** [comment]
@nmco conflicts removed and tests passed, ready for merge please

File: src/main/src/main/java/org/geoserver/catalog/WMSLayerInfo.java [hunk 10]
```diff
@@ -55,5 +55,9 @@
 
     public StyleInfo getDefaultStyle();
 
+    public boolean isRespectMetadataBBox();
```
Comment (aaime) [inline src/main/src/main/java/org/geoserver/catalog/WMSLayerInfo.java]: Just a quick thing, can you make this "english readable"? IsMetadataBBoxRespected()?

Comment (imranrajjad) [reply src/main/src/main/java/org/geoserver/catalog/WMSLayerInfo.java]: sure

Comment (imranrajjad) [reply src/main/src/main/java/org/geoserver/catalog/WMSLayerInfo.java]: @aaime requested changes applied

**Comment by nmco** [comment]
Looks fine I'm merging, thx @imranrajjad. Note, didn't test this manually.

File: doc/en/user/source/data/cascaded/wms.rst [hunk 11]
```diff
@@ -62,6 +62,8 @@ Connecting a remote WMS allows for the following features:
 
 * Full **REST Configuration**. See the :ref:`rest` section for more information about the GeoServer REST interface.
 
+* **Optimization**. It is possible to ignore remote WMS calls if requested map area is completely outside the native bounds as advertised by remote WMS layer. These settings are available under **Publishing** Tab.
+
```
Comment (jodygarnett) [inline doc/en/user/source/data/cascaded/wms.rst]: This documentation update was insufficient:

* REST API is documented using OpenAPI
* User interface changes to the publishing tab should be documented

Comment (imranrajjad) [reply doc/en/user/source/data/cascaded/wms.rst]: @jodygarnett not sure about the context of REST API.

for the optimization part, will surely add images to documentation

Comment (jodygarnett) [reply doc/en/user/source/data/cascaded/wms.rst]: The rest api stuff is [here](https://github.com/geoserver/geoserver/tree/master/doc/en/api/1.0.0) - specially this: https://docs.geoserver.org/latest/en/api/#/latest/en/api/1.0.0/wmsstores.yaml

Thanks for updating the docs!
